### PR TITLE
Using Metal allows training on the GPU of Apple devices 

### PR DESCRIPTION
**Pull request by gklab**
Set the parameter device to "mps" to train on Apple devices. Tested on the M2 Ultra chip, the training speed is 6-7 times that of "cpu".

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -27,10 +27,10 @@
 from model import Transformer, ModelArgs
 from torch.distributed import destroy_process_group, init_process_group
 from torch.nn.parallel import DistributedDataParallel as DDP
-
+import torch._dynamo
 from tinystories import Task
 from export import model_export
-
+torch._dynamo.config.suppress_errors = True
```
Comment (karpathy) [inline train.py:33]: not familiar with these, why did you add these two lines? are they safe? importing _modules feels wrong

Comment (gklab) [reply train.py]: > not familiar with these, why did you add these two lines? are they safe? importing _modules feels wrong

Yes, this approach is extreme.
I just adjusted to turn off compilation optimization under MPS. Because the compilation of torch is not complete for MPS, it currently supports backend="aot_eager", but the speed is a little slower than not compiling.

Comment (gklab) [reply train.py]: > not familiar with these, why did you add these two lines? are they safe? importing _modules feels wrong

In addition, the reason why adding these two lines can make mps run normally is because mps will report an error when compiling, and then ignore the compilation behavior. That is to say, mps training is not compiled.

**Comment by karpathy** [comment]
I do want to merge this PR but why is the formatitng so weird? There are many spurious changes here - prints get split up to multiple lines, new lines appear or disappear? If you're able to clean this up to a minimal change that introduces the desired mps behavior without touching anything else I'd be happy to merge.

**Comment by gklab** [comment]
> I do want to merge this PR but why is the formatitng so weird? There are many spurious changes here - prints get split up to multiple lines, new lines appear or disappear? If you're able to clean this up to a minimal change that introduces the desired mps behavior without touching anything else I'd be happy to merge.

Oh, it may be caused by my habitual line wrapping, and part of the reason is the automatic formatting of the IDE.
But this does cause problems for collaborative programming.
I have deleted the redundant code and hope to PR smoothly.

**Comment by gklab** [comment]
> I do want to merge this PR but why is the formatitng so weird? There are many spurious changes here - prints get split up to multiple lines, new lines appear or disappear? If you're able to clean this up to a minimal change that introduces the desired mps behavior without touching anything else I'd be happy to merge.

I've made another adjustment and it shouldn't be simpler.
